cachito/workers/pkg_managers/gomod.py

```diff
@@ -91,38 +94,118 @@ class GoPackage(_GolangModel):
 
-    Cachito will reuse the same cache directory between retries, so Go will not have to download
-    the same dependency twice. The backoff is exponential, Cachito will wait 1s -> 2s -> 4s -> ...
-    before retrying.
+    Provides convenient methods to download project dependencies, alternative toolchains,
+    parses various Go files, etc.
     """
-    n_tries = get_worker_config().cachito_gomod_download_max_tries
 
-    @backoff.on_exception(
-        backoff.expo,
-        CachitoCalledProcessError,
-        jitter=None,  # use deterministic backoff, do not apply jitter
-        max_tries=n_tries,
-        logger=log,
-    )
-    def run_go(_cmd, _params) -> str:
-        log.debug(f"Running {_cmd}")
-        return run_gomod_cmd(_cmd, _params)
-
-    try:
-        return run_go(cmd, params)
-    except CachitoCalledProcessError:
-        err_msg = (
-            f"Processing gomod dependencies failed. Cachito tried the {' '.join(cmd)} command "
-            f"{n_tries} times. This may indicate a problem with your repository or Cachito itself."
+    def __init__(
+        self,
+        binary: Union[str, os.PathLike[str]] = "go",
+        release: Optional[str] = None,
+    ) -> None:
+        """Initialize the Go toolchain wrapper.
+
+        :param binary: path-like string to the Go binary or direct command (in PATH)
+        :param release: Go release version string, e.g. go1.20, go1.21.10
+        :returns: a callable instance
+        """
+        # run_cmd will take care of checking any bogus passed in 'binary'
+        self._bin = str(binary)
+        self._release = release
+
+        if self._release:
+            self._bin = f"/usr/local/go/{release}/bin/go"
+
+        self._version: Optional[pkgver.Version] = None
+
+    @tracer.start_as_current_span("Go.__call__")
+    def __call__(self, cmd: list[str], params: Optional[dict] = None, retry: bool = False) -> str:
+        """Run a Go command using the underlying toolchain, same as running GoToolchain()().
+
+        :param cmd: Go CLI options
+        :param params: additional subprocess arguments, e.g. 'env'
+        :param retry: whether the command should be retried on failure (e.g. network actions)
+        :returns: Go command's output
+        """
+        if params is None:
+            params = {}
+
+        cmd = [self._bin] + cmd
+        if retry:
+            return self._retry(cmd, **params)
+
+        return self._run(cmd, **params)
+
+    @property
+    def version(self) -> pkgver.Version:
+        """Version of the Go toolchain as a packaging.version.Version object."""
+        if not self._version:
+            self._version = pkgver.Version(self.release[2:])
+        return self._version
+
+    @property
+    def release(self) -> str:
+        """Release name of the Go Toolchain, e.g. go1.20 ."""
+        # lazy evaluation: defer running 'go'
+        if not self._release:
+            output = self(["version"])
+            log.info(f"Go release: {output}")
```

Comment:
This will usually generate a log very similar to what this [line](https://github.com/containerbuildsystem/cachito/blob/master/cachito/workers/tasks/gomod.py#L133) does:

```
INFO Go version: go version go1.20.13 linux/amd64
(...)
INFO Running '['go', 'version']'
INFO Go release: go version go1.20.13 linux/amd64
INFO go.mod recommends/requires Go version: 1.21
(...)
```

I think it would be a little clearer if the logs were something such as:
```
INFO go.mod recommends/requires Go version: 1.21
INFO Processing request using go version go1.21.0 linux/amd64
```

Or just remove the original line in the tasks.gomod.